transforms/universal/fdedup/python/src/cluster_analysis_transform.py

```diff
@@ -0,0 +1,336 @@
+        """
+        super().__init__(config)
+        self.num_bands = config.get(num_bands_key, num_bands_default)
+        self.num_segments = config.get(num_segments_key, num_segments_default)
+        self.jaccard_similarity_threshold = config.get(
+            jaccard_similarity_threshold_key, jaccard_similarity_threshold_default
+        )
+        self.sort_output = config.get(sort_output_key, sort_output_default)
+        self.data_access = config.get("data_access")
+        self.logger = get_logger(__name__)
+
+    def transform(self, folder_name: str) -> tuple[list[tuple[bytes, str]], dict[str, Any]]:
+        self.logger.info(f"Cluster analysis for folder {folder_name}")
+        metadata = {}
+        input_folder = self.sanitize_folder_name(os.path.join(self.data_access.input_folder, folder_name))
+        files, retries = self.data_access.get_folder_files(
+            path=input_folder,
+            extensions=[".parquet"],
+            return_data=True,
+        )
+        if retries > 0:
+            metadata |= {"data_access_retries": retries}
+        match = re.match(r"^band=(\d+)/segment=(\d+)$", folder_name)
+        if match:
+            band = int(match.group(1))
+            segment = int(match.group(2))
+        else:
+            raise ValueError(f"Wrong folder_name {folder_name}, should be band=b/segment=s")
+        output_folder = self.sanitize_folder_name(self.data_access.output_folder)
+        output_path = os.path.join(output_folder, f"band_{band}_segment_{segment}.parquet")
+
+        # consolidate into a single data frame band hashes computed by workers
+        band_segment_dataframe, consolidation_stats = self.consolidate_band_segment_files(files)
+        metadata |= consolidation_stats
+        # cluster grouping by band hashes
+        cluster_dataframe, cluster_stats = self.get_clusters(band_segment_dataframe)
+        metadata |= cluster_stats
+        # cluster analysis using jaccard similarity
+        jaccard_cluster_dataframe, jaccard_stats = self.analyze_clusters(cluster_dataframe)
+        metadata |= jaccard_stats
+        # Generate the docs_to_remove dataframe
+        docs_to_remove_dataframe = jaccard_cluster_dataframe.explode("docs_to_remove")
+        output_data = TransformUtils.convert_arrow_to_binary(docs_to_remove_dataframe.to_arrow())
+        self.logger.info(f"{len(docs_to_remove_dataframe)} documents marked to remove")
+        metadata |= {"num_duplicate_documents": len(docs_to_remove_dataframe)}
+        return [(output_data, output_path)], metadata
+
+    def sanitize_folder_name(self, folder_name: str) -> str:
+        if "://" in folder_name:
+            _, folder_name = folder_name.split("://")
+        if folder_name[-1] != "/":
+            folder_name = f"{folder_name}/"
+        return folder_name
+
+    def consolidate_band_segment_files(self, files: dict[str, bytes]) -> tuple[pl.DataFrame, dict[str, Any]]:
```

Comment:
I think you should hide all these internal methods by prefixing them with `_` or `__`.

Reply:
Fixed in commit https://github.com/IBM/data-prep-kit/pull/699/commits/8de14b51e55a69bfa7f4dddc763500fb3a01db4d